main.go

```diff
@@ -178,90 +193,213 @@ func main() {
 		})
 	}
 	{
+		l := log.With(l, "component", "writer")
 		t := time.NewTicker(opts.Period)
 		ctx, cancel := context.WithCancel(ctx)
 
+		f := func() {
+			if err := write(ctx, opts.WriteEndpoint, opts.Token, generate(opts.Labels), l); err != nil {
+				m.remoteWriteRequests.WithLabelValues("error").Inc()
+				level.Error(l).Log("msg", "failed to make request", "err", err)
+			} else {
+				m.remoteWriteRequests.WithLabelValues("success").Inc()
+			}
+		}
+
 		g.Add(func() error {
-			level.Info(logger).Log("msg", "starting the remote-write client")
+			level.Info(l).Log("msg", "starting the remote-write client")
+			// Write first metric immediately after start without any delay.
+			f()
 			for {
 				select {
 				case <-t.C:
-					if err := write(ctx, endpointWrite, opts.Token, generate(opts.Labels)); err != nil {
-						level.Error(logger).Log("msg", "failed to make request", "err", err)
-					}
+					f()
 				case <-ctx.Done():
-					return nil
+					return reportResults(l, m.remoteWriteRequests, opts.SuccessThreshold)
 				}
 			}
 		}, func(_ error) {
 			t.Stop()
 			cancel()
 		})
 	}
-	{
-		var cancel func()
 
+	if opts.ReadEndpoint != nil {
+		l := log.With(l, "component", "reader")
+		t := time.NewTicker(opts.Period)
+
+		var cancel func()
 		if opts.Duration != 0 {
 			ctx, cancel = context.WithTimeout(ctx, opts.Duration)
 		} else {
 			ctx, cancel = context.WithCancel(ctx)
 		}
 
-		t := time.NewTicker(opts.Period)
+		g.Add(func() error {
+			level.Info(l).Log("msg", "waiting for initial delay before querying for metrics")
+			// Wait for at least one period before start reading metrics.
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(opts.InitialQueryDelay + opts.Period):
```

Comment:
I don't think this quite makes sense now. You implemented both of the changes, where only one should have been made. Now, there will be two writes before a single read is attempted:
t=0 metric is written
t=1 timers start
t=period metric is written
t=period+delay metric is read

it should EITHER be:
t=0 metric is written
t=1 timers start
_t=delay metric is read_
t=period metric is written
t=period+delay metric is read

OR:
t=1 timers start
t=period metric is written
t=period+delay metric is read

From these two options, I think the last one is simpler and delay can be any value

Reply:
@squat Alright, I went with the simpler option. So now it starts writing after `t=period` and starts reading after `t=period+delay`.